opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py

```diff
@@ -377,6 +377,12 @@ def __init__(
         end_time: Optional[int] = None,
         instrumentation_scope: Optional[InstrumentationScope] = None,
     ) -> None:
+        if not isinstance(name, str):
+            stack_trace = "".join(traceback.format_stack())
+            logger.warning(
+                "span name must be a string.\n Stack trace: %s", stack_trace
+            )
+            name = str(name)
```

Comment:
Did you consider cases where converting the `name` to a string might fail?

Reply:
I cannot really think of any case where this should fail. Did you have any such cases in mind?

Reply:
It's good that we're logging something, but I'm not sure that we want to log a stack trace here. Is logging a stacktrace done elsewhere in this repo?

Regarding testing, I think we should test passing in a string, and test passing in a non-string, making sure both scenarios produce expected results.

Reply:
I don't see logging a stacktrace being done anywhere in the repo. Would you like me to change it and if so, how would you like me to change it?

Sure, working on it now.

Reply:
Maybe this should just raise a TypeError. Anybody have thoughts?

```
        if not isinstance(name, str):
            raise TypeError(f"Expected 'name' to be a string, but got {type(name).__name__}")
```

Reply:
@pmcollins, and in that case, we don't try to convert to str, right?
  I was wondering about [this](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/error-handling.md#basic-error-handling-principles): 
>API methods MUST NOT throw unhandled exceptions when used incorrectly by end users. The API and SDK SHOULD provide safe defaults for missing or invalid arguments. For instance, a name like empty may be used if the user passes in null as the span name argument during Span construction.

I didn't find a clarification on what is a valid span name other than this: 
>The span name SHOULD be the most general string that identifies a (statistically) interesting class of Spans rather than individual Span instances while still being human-readable. That is, "get_user" is a reasonable name, while "get_user/314159", where "314159" is a user ID, is not a good name due to its high cardinality. Generality SHOULD be prioritized over human-readability.

Reply:
Yes, good point, the exception would have to be handled.

Reply:
Sure, will do that. Also, should I just raise a TypeError as you stated before or should I convert it to a string and continue execution like it is now?

Reply:
I think we shouldn't convert to str and we must also be careful not to generate excessive warnings.

Reply:
@soumyadeepm04 

We didn't happen to catch your for 07/18's Python SIG. Feel free to bring this issue up in next week's meeting.

Reply:
We discussed this in SIG 07/25. Since we didn't decide on the solution we should adopt, we need to continue the discussions here.

Reply:
Seems like there's two thing we can do:

1. Treat this as a misconfiguration/misuse of the api (even though it's not explictly stated that span name is a string) similar to [here](https://github.com/open-telemetry/opentelemetry-python/blob/main/opentelemetry-sdk/src/opentelemetry/sdk/_logs/_internal/export/__init__.py#L452) and raise a typeerror in the create span api
2. Handle this gracefully and log a warning instead -> attempt to convert the value to a str in the api and throw an error if this attempt fails.

I personally am more for (1), this seems more like a misuse of an api in which every we expect name to be a string. Otherwise, we must do type checks every we use span name.

Reply:
Raising an error here is against the OTel [basic error handling principles](https://opentelemetry.io/docs/specs/otel/error-handling/#basic-error-handling-principles) i believe. Since `ReadableSpan.__init__` is called on `Span.end` a raised error will escape the API/SDK and fail the user's application.
Converting to str is probably fine but should be handled with care since e.g. a custom `__str__` implementation might also raise an error.
Another option might be to just replace the span name with some hardcoded "i'm not a str pls fix me" message so users would see in their backend that something isn't quite right.

Reply:
Given the specification quotes here https://github.com/open-telemetry/opentelemetry-python/pull/3955#discussion_r1633547724 and that we have the proper type on the method parameter I'll just leave the code as is. Hopefully in the future users would be able to catch these with a type checker?

Reply:
@xrmx 

Should we close the pr in this case? @mariojonke  wdyt

Reply:
> Should we close the pr in this case? @mariojonke wdyt

Yes